concurrency/ConnectionMultiplexerManager.java

```diff
@@ -0,0 +1,385 @@
+        // Start thread that will send heartbeats to Connection Managers every 30 seconds
+        // to keep connections open.
+        TimerTask heartbeatTask = new TimerTask() {
+            @Override
+            public void run() {
+                try {
+                    for (ConnectionMultiplexerSession session : sessionManager.getConnectionMultiplexerSessions()) {
+                        session.deliverRawText(" ");
+                    }
+                }
+                catch(Exception e) {
+                    Log.error(e.getMessage(), e);
+                }
+            }
+        };
+        TaskEngine.getInstance().schedule(heartbeatTask, 30*JiveConstants.SECOND, 30*JiveConstants.SECOND);
+    }
+
+    /**
+     * Creates a new client session that was established to the specified connection manager.
+     * The new session will not be findable through its stream ID.
+     *
+     * @param connectionManagerDomain the connection manager that is handling the connection
+     *        of the session.
+     * @param streamID the stream ID created by the connection manager for the new session.
+     * @param hostName the address's hostname of the client or null if using old connection manager.
+     * @param hostAddress the textual representation of the address of the client or null if using old CM.
+     * @return true if a session was created or false if the client should disconnect.
+     */
+    public boolean createClientSession(String connectionManagerDomain, String streamID, String hostName, String hostAddress) {
+        Connection connection = new ClientSessionConnection(connectionManagerDomain, hostName, hostAddress);
+        // Check if client is allowed to connect from the specified IP address. Ignore the checking if connection
+        // manager is old version and is not passing client's address
+        byte[] address = null;
+        try {
+            address = connection.getAddress();
+        } catch (UnknownHostException e) {
+            // Ignore
+        }
+        if (address == null || LocalClientSession.isAllowed(connection)) {
+            LocalClientSession session =
+                    SessionManager.getInstance().createClientSession(connection, new BasicStreamID(streamID));
+            // Register that this streamID belongs to the specified connection manager
+            streamIDs.put(streamID, connectionManagerDomain);
+            // Register which sessions are being hosted by the speicifed connection manager
+            /*** BEGIN amazon modifications (DCL Security fix ***/
+            Map<String, LocalClientSession> sessions;
+            synchronized (connectionManagerDomain.intern()) {
+                sessions = sessionsByManager.get(connectionManagerDomain);
+                if (sessions == null) {
+                    sessions = new ConcurrentHashMap<>();
+                    sessionsByManager.put(connectionManagerDomain, sessions);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **167 and 170**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `put()` with `putIfAbsent()` and then removing the synchronization. The check is at line: **168**.

**More info**  
 `putIfAbsent()` is thread-safe and puts the value only if the `ConcurrentHashMap` does not contain the key. `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-mapreduce-project/hadoop-mapreduce-client/hadoop-mapreduce-client-hs/src/main/java/org/apache/hadoop/mapreduce/v2/hs/HistoryFileManager.java#L927-L932) (external link).